### DISCUSSION TITLE
Simple RPM value instead of counter for CPU PWM Fans. ( Or Generalized tachometer)

### DISCUSSION THREAD

**Discussion by thomasesr** [arendst/Tasmota#11257] [Ideas]
I have two CPU PWM fans that I want to control using the Tasmota, I've managed to get them working fine by changing the max PWM Frequency allowed from the hardcoded 4000 to 25000 (I don't know why this 4000hz limitation, but it seam to work fine setting it to 25khz on the Nidec 3.3A Fans I'm using to move air outside my server rack). 

I've also connected the tachometer output of the fans into different IO pins on the ESP8266 and if I set them to Counter on the Tasmota Configuration I can see them counting up no problem when their moving. However I would like for the MCU to display the information as RPM instead of giving the raw count.

It should be simple enough to implement (even though I cannot do it in a simple elegant manner) and should be more precise than trying to calculate from mqtt. Also less taxing on heap memory since you don't need to store a huge value like the counter does (specially if you have a 5000 RPM fan).

**Comment by thomasesr** [comment]
[The 4-Pin PWM Fan specification paper from Intel Rev 1.3 September 2005](https://wiki.kobol.io/helios4/files/fan/4_Wire_PWM_Spec.pdf)

The specification states 2 pulses per revolution. So a 5000 RPM Fan (very fast) should give you 166 pulses per second while 1000 RPM (normal to slow) should be about 33.3 pulses per second. each second should be enough pulses/data to calculate the RPM. Some basic math and logic:

Configurable Settings:
**PPR** (pulses per revolution): 2
**STF** (sampled time frame): 2000ms
**FTT** (frequency time table): either 1,000ms for RPS (revolutions per second) or 60,000ms for RPM (revolutions per minute)

Calculate the value of the settings once into a variable on boot up, so if you change the settings you would have to restart the MCU (lets call it **Y**):

**Y** = **FTT/STF/PPR** (60000/2000/2)
so **Y** = 15

Loop Logic:
Sample **PC** (Pulse counter) for the configured **STF**)
Calculate RPM/RPS: **PC** * **Y** 
Reset Pulse Counter
Repeat

(for example, RPM should be the counted pulses in the time frame in milliseconds multiplied by 60seconds in milliseconds divided by the sampled time frame time in ms divided by the PPR or Y)

So: 235 * 60000/2000/2
or 235 * 15 = 3525 RPM

You can hardcode the settings values into the code, but it would be nice to be able to change this values from the console/mqtt to configure a faster refresh rate on the fan speeds or even use a different motor/equipment all together and make a generalized tachometer. If you have 1 PPR or 4 PPR or want to set a 500ms refresh rate on the speed of a super fast motor or even get rotations per second.

**Comment by lalo-uy** [comment]
You can have a Timer rule that substracts the previous count , do the
scaling and send the mqtt.
I belive the actual count can not be read, but must be saved to a var with
a on counter rule.

**Comment by lalo-uy** [comment]
Something like this

Rule1 ON counter#c1 do var1 %value%; ENDONE ON TIME#MINUTE do backlog var3
%var1%;  sub3 %var2% ; var2 %var1%;  publish tele/fan/speed %var3%
